pennylane/devices/preprocess.py

```diff
@@ -147,6 +147,19 @@ def validate_device_wires(
     return (tape,), null_postprocessing
 
 
+@transform
+def mid_circuit_measurements(
+    tape: qml.tape.QuantumTape, device
+) -> (Sequence[qml.tape.QuantumTape], Callable):
+    """Provide the transform to handle mid-circuit measurements.
+
+    If the tape of device uses finite-shot, use the native implementation (i.e. no transform), and use defer measurements transforms otherwise.
+    """
+    if tape.shots:
+        return (tape,), null_postprocessing
+    return qml.defer_measurements(tape, device=device)
```

Comment:
```suggestion
    if tape.shots and (tape.batch_size is None or not any(isinstance(o, MidMeasureMP) for o in tape.operations)):
        return (tape,), null_postprocessing
    return qml.defer_measurements(tape, device=device)
```

Since native mid circuit measurements are not compatible with parameter broadcasting, we should still apply `defer_measurements` if there's a batch dimension.

Strictly applying `qml.transforms.broadcast_expand` doesn't quite seem to work in this case, so we should just fallback to `defer_measurements`.

Reply:
Good point, I'll take your change.